### PR TITLE
Fix the regex to match client secret in OAuth requests

### PR DESCRIPTION
**Pull request by Jinming-Hu**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/core/azure-core-test/src/test_proxy_manager.cpp [hunk 1]
```diff
@@ -242,7 +242,7 @@ void TestProxyManager::SetProxySanitizer()
 
   addSanitizer(SanitizerType::General, g_accountRegex, "account");
   addSanitizer(
-      SanitizerType::Body, "client_secret=(?<clientsecret>[a-zA-Z0-9\\%]+)", "clientsecret");
+      SanitizerType::Body, "client_secret=(?<clientsecret>[a-zA-Z0-9\\%_~]+)", "clientsecret");
```
Comment (antkmsft) [inline sdk/core/azure-core-test/src/test_proxy_manager.cpp:245]:
```suggestion
      SanitizerType::Body, "client_secret=(?<clientsecret>[a-zA-Z0-9\\%_~\-\.]+)", "clientsecret");
```
This should fix it once and for all.

Explanation:
https://github.com/Azure/azure-sdk-for-cpp/blob/15d8133a215a1f1d93c40714455e9f4e8afb4013/sdk/identity/azure-identity/src/client_secret_credential.cpp#L32

=>

https://github.com/Azure/azure-sdk-for-cpp/blob/15d8133a215a1f1d93c40714455e9f4e8afb4013/sdk/core/azure-core/src/http/url.cpp#L134

And, I just made Azure Portal to generate 5 random AAD secrets for me, some of them had a dash and a dot.

**Review by antkmsft — APPROVED** [review]

**Comment by Jinming-Hu** [comment]
@antkmsft Here's an update.

I think the test failures was caused by a change in https://github.com/Azure/azure-sdk-for-cpp/pull/5051, where we enabled record-playback testing for WinHTTP transport.

After some investigation, I found the failures were caused by the differences between libcurl and WinHTTP.

1. WinHTTP doesn't send `Cookie` header when getting OAuth tokens.
1. WinHTTP doesn't send `ContentLength: 0` HTTP request header for `PATCH` requests if the body is empty.

In this PR, we fixed the first problem by ignoring `Cookie` header when matching requests with recordings.

The second problem can be fixed with patch below, but it involves modifications to win http transport. I'd like to leave the decision on how to fix it up to your team 

```PATCH
diff --git a/sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp b/sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp
index 199d741f..6841064c 100644
--- a/sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp
+++ b/sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp
@@ -209,6 +209,11 @@ std::string GetHeadersAsString(Azure::Core::Http::Request const& request)
     requestHeaderString += header.second; // string's value
     requestHeaderString += "\r\n";
   }
+  if (request.GetMethod() == Azure::Core::Http::HttpMethod::Patch
+      && request.GetHeaders().count("Content-Length") == 0)
+  {
+    requestHeaderString += "Content-Length: 0\r\n";
+  }
   requestHeaderString += "\r\n";

   return requestHeaderString;
```

**Comment by LarryOsterman** [comment]
> @antkmsft Here's an update.
> 
> I think the test failures was caused by a change in #5051, where we enabled record-playback testing for WinHTTP transport.
> 
> After some investigation, I found the failures were caused by the differences between libcurl and WinHTTP.
> 
> 1. WinHTTP doesn't send `Cookie` header when getting OAuth tokens.
> 2. WinHTTP doesn't send `ContentLength: 0` HTTP request header for `PATCH` requests if the body is empty.
> 
> In this PR, we fixed the first problem by ignoring `Cookie` header when matching requests with recordings.
> 
> The second problem cannot be fixed with test-proxy directives. Test-proxy framework is just not smart enough to treat 0-content-length and no Content-Length header at all as equals. It can be fixed with patch below, but it involves modifications to win http transport. I'd like to leave the decision on how to fix it up to your team
> 
> ```diff
> diff --git a/sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp b/sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp
> index 199d741f..6841064c 100644
> --- a/sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp
> +++ b/sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp
> @@ -209,6 +209,11 @@ std::string GetHeadersAsString(Azure::Core::Http::Request const& request)
>      requestHeaderString += header.second; // string's value
>      requestHeaderString += "\r\n";
>    }
> +  if (request.GetMethod() == Azure::Core::Http::HttpMethod::Patch
> +      && request.GetHeaders().count("Content-Length") == 0)
> +  {
> +    requestHeaderString += "Content-Length: 0\r\n";
> +  }
>    requestHeaderString += "\r\n";
> 
>    return requestHeaderString;
> ```

FWIW, I concur with this analysis - I came to the same conclusion after looking at the logs from the failure last night. The test proxy infrastructure is biased towards cURL if both CURL and WinHTTP are available, which is why this issue went unnoticed.

**Review by LarryOsterman — APPROVED** [review]

**Comment by Jinming-Hu** [comment]
I'll merge this PR first. This doesn't affect #5071 because they fix different problems.

**Pull request merged**